src/OpenLoco/src/World/TownManager.cpp

```diff
@@ -29,12 +34,140 @@ namespace OpenLoco::TownManager
     }
 
+    // 0x00497D70
+    static uint32_t copyTownNameToBuffer(const TownNamesObject* namesObj, uint32_t offset, uint16_t index, char* buffer)
+    {
+        auto* offsetPtr = reinterpret_cast<const std::byte*>(namesObj) + offset;
+        auto srcOffset = *reinterpret_cast<const int16_t*>(offsetPtr + index * 2);
+
+        auto* srcPtr = reinterpret_cast<const char*>(offsetPtr + srcOffset);
+        strcpy(buffer, srcPtr);
+
+        uint8_t tail = *reinterpret_cast<const uint8_t*>(srcPtr + strlen(srcPtr) + 1);
+        return tail;
+    }
+
+    // 0x00497A6A
+    static uint8_t townNameFromNamesObject(uint32_t& rand, const char* buffer)
+    {
+        auto* namesObj = ObjectManager::get<TownNamesObject>();
+        uint8_t testsToRun = 0;
+
+        for (auto& category : namesObj->categories)
+        {
+            if (category.count == 0)
+            {
+                continue;
+            }
+
+            uint16_t ax = rand;
+            uint16_t dx = category.count + category.bias;
+            int16_t index = ((ax * dx) >> 16) - category.bias;
+
+            if (index > 0)
+            {
+                char* strEnd = const_cast<char*>(buffer + strlen(buffer));
+                testsToRun |= copyTownNameToBuffer(namesObj, category.offset, index, strEnd);
+            }
+
+            for (auto shifts = category.count + category.bias; shifts > 0; shifts >>= 1)
+            {
+                rand = std::rotr(rand, 1);
+            }
+        }
+
+        return testsToRun;
+    }
+
     // 0x004978B7
     static bool generateTownName(Town* town)
     {
-        registers regs;
-        regs.esi = X86Pointer(town);
-        return !(call(0x004978B7, regs) & X86_FLAG_CARRY);
+        for (auto attemptsLeft = 400U; attemptsLeft > 0; attemptsLeft--)
+        {
+            char buffer[256]{};
+            auto rand = town->prng.randNext();
+            auto testsToRun = townNameFromNamesObject(rand, buffer);
+
+            if (strlen(buffer) == 0)
+            {
+                continue;
+            }
+
+            if (strlen(buffer) > StringManager::kUserStringSize)
+            {
+                continue;
+            }
+
+            if (Gfx::TextRenderer::getStringWidth(Gfx::Font::medium_bold, buffer) > 200)
+            {
+                continue;
+            }
+
+            // clang-format off
+            if (testsToRun & 1)
+            {
+                // Check that the town is adjacent to a large amount of water tiles on at least one side.
+                auto pos = Pos2(town->x, town->y);
+                if (!(TileManager::countSurroundingWaterTiles(pos + Pos2(6 * kTileSize, 0)) > 65 ||
+                    TileManager::countSurroundingWaterTiles(pos + Pos2(0, 6 * kTileSize)) > 65 ||
+                    TileManager::countSurroundingWaterTiles(pos + Pos2(0 - 6 * kTileSize, 0)) > 65 ||
+                    TileManager::countSurroundingWaterTiles(pos + Pos2(0, 0 - 6 * kTileSize)) > 65))
```

Comment:
Perhaps we should make a helper function for these calls, with a threshold parameter (here: `65`, and `15` later later on in this function) .

Reply:
Yes, only useful for this function but would make things easier to read.